src/plugins/intel_gpu/src/kernel_selector/cl_kernels/sdpa_opt.cl

```diff
@@ -1167,63 +1166,52 @@ KERNEL(sdpa_opt)(
                     qk_acc[i] = INPUT0_MIN_FUNC(INPUT0_MAX_FUNC(qk_acc[i], INPUT0_VAL_MIN), INPUT0_VAL_MAX);
 
                     qk_max = SOFTMAX_ACCUMULATOR_MAX_FUNC(qk_max, TO_SOFTMAX_ACCUMULATOR_TYPE(qk_acc[i]));
+                    slm_qk_vals[sglid][sgid * TARGET_SEQ_LEN_BLOCK_SIZE + i] = qk_acc[i];
                 }
-            }
-
-            {
-                slm_qk_max_vals[sgid * SUBGROUP_SIZE + sglid] = qk_max;
-                qk_max = SOFTMAX_ACCUMULATOR_VAL_MIN;
+                slm_qk_max_vals[sglid][sgid] = qk_max;
             }
 
         barrier(CLK_LOCAL_MEM_FENCE);
 
         {
             // SoftMax calculation
-            SOFTMAX_ACCUMULATOR_TYPE qk_max_new = SOFTMAX_ACCUMULATOR_VAL_MIN;
-
-            for (uint i = 0; i < SUBGROUPS_PER_WG; i++) {
-                SOFTMAX_ACCUMULATOR_TYPE qk_max_val = slm_qk_max_vals[i * SUBGROUP_SIZE + sglid];
-                qk_max_new = SOFTMAX_ACCUMULATOR_MAX_FUNC(qk_max_new, qk_max_val);
-            }
-
-            if (sgid == 0) {
-                slm_max_val_cur[sglid] = qk_max_new;
-            }
-
-            SOFTMAX_ACCUMULATOR_TYPE exp_sum_new = SOFTMAX_ACCUMULATOR_VAL_ZERO;
-
-            for (uint i = 0; i < TARGET_SEQ_LEN_BLOCK_SIZE; i++) {
-                qk_acc[i] = native_exp(TO_SOFTMAX_ACCUMULATOR_TYPE(qk_acc[i]) - qk_max_new);
-                exp_sum_new += qk_acc[i];
-            }
-
-            {
-                slm_exp_sum_vals[sgid * SUBGROUP_SIZE + sglid] = exp_sum_new;
-            }
-
-            exp_sum_new = SOFTMAX_ACCUMULATOR_VAL_ZERO;
-
-            barrier(CLK_LOCAL_MEM_FENCE);
-
-            for (uint i = 0; i < SUBGROUPS_PER_WG; i++) {
-                SOFTMAX_ACCUMULATOR_TYPE exp_sum = slm_exp_sum_vals[i * SUBGROUP_SIZE + sglid];
-                exp_sum_new += exp_sum;
-            }
-
-            for (uint i = 0; i < TARGET_SEQ_LEN_BLOCK_SIZE; i++) {
-                qk_acc[i] = qk_acc[i] / exp_sum_new;
-            }
-
-            if (sgid == 0) {
-                slm_exp_sum_cur[sglid] = exp_sum_new;
-            }
-
-            for (uint i = 0; i < TARGET_SEQ_LEN_BLOCK_SIZE; i++) {
-                slm_qk_vals[sglid * SEQ_LEN_PARTITION_SIZE + sgid * TARGET_SEQ_LEN_BLOCK_SIZE + i] = qk_acc[i];
+            // each sg will compute a whole row of query
+            const int key_len_in_kv_block = SEQ_LEN_PARTITION_SIZE;
+            for (uint m = sgid; m < seq_idx_end; m += SUBGROUPS_PER_WG) {
+                // rowmax
+                SOFTMAX_ACCUMULATOR_TYPE qk_max_new;
+                if (sglid < SUBGROUPS_PER_WG) { // TODO: if SUBGROUPS_PER_WG > 16?
```

Comment:
What happen if SUBGROUPS_PER_WG > 16?

Reply:
This is not expected to happen. The maximum allowed head_size is 256, so SUBGROUPS_PER_WG shouldn't be higher than 16.:
https://github.com/openvinotoolkit/openvino/blob/357eb5415891034e7b10895f72e736752e29975a/src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp#L444
This should also be taken into account in this function, but it was missed:
https://github.com/openvinotoolkit/openvino/blob/357eb5415891034e7b10895f72e736752e29975a/src/plugins/intel_gpu/src/kernel_selector/kernels/sdpa/sdpa_kernel_opt.cpp#L25